examples/step-98/step-98.cc

```diff
@@ -0,0 +1,974 @@
+        level_constraints.reinit(relevant_dofs);
+        level_constraints.add_lines(
+          mg_constrained_dofs.get_boundary_indices(level));
+        level_constraints.close();
+
+        typename MatrixFree<dim, LevelNumber>::AdditionalData additional_data;
+        additional_data.tasks_parallel_scheme =
+          MatrixFree<dim, LevelNumber>::AdditionalData::none;
+        additional_data.mapping_update_flags =
+          (update_gradients | update_JxW_values | update_quadrature_points);
+        additional_data.mg_level = level;
+
+        // We need to set the level of the matrix-free object to
+        // enable evaluation on ghosted cell.
+        additional_data.store_ghost_cells = true;
+        std::shared_ptr<MatrixFree<dim, LevelNumber>> mg_mf_storage_level(
+          new MatrixFree<dim, LevelNumber>());
+        mg_mf_storage_level->reinit(mapping,
+                                    dof_handler,
+                                    level_constraints,
+                                    QGauss<1>(fe.degree + 1),
+                                    additional_data);
+
+        mg_matrices[level].initialize(mg_mf_storage_level,
+                                      mg_constrained_dofs,
+                                      level);
+      }
+    setup_time += time.wall_time();
+    time_details << "Setup matrix-free levels   (CPU/wall) " << time.cpu_time()
+                 << "s/" << time.wall_time() << "s" << std::endl;
+  }
+
+
+
+  template <int dim, int fe_degree>
+  void LaplaceSolver<dim, fe_degree>::assemble_rhs()
+  {
+    Timer time;
+
+    system_rhs = 0;
+    FEEvaluation<dim, fe_degree> phi(*system_matrix.get_matrix_free());
+    for (unsigned int cell = 0;
+         cell < system_matrix.get_matrix_free()->n_cell_batches();
+         ++cell)
+      {
+        phi.reinit(cell);
+        for (unsigned int q = 0; q < phi.n_q_points; ++q)
+          phi.submit_value(make_vectorized_array<Number>(1.0), q);
+        phi.integrate(EvaluationFlags::values);
+        phi.distribute_local_to_global(system_rhs);
+      }
+    system_rhs.compress(VectorOperation::add);
+
+    setup_time += time.wall_time();
+    time_details << "Assemble right hand side   (CPU/wall) " << time.cpu_time()
+                 << "s/" << time.wall_time() << "s" << std::endl;
+  }
+
+  template <int dim, int fe_degree>
+  void LaplaceSolver<dim, fe_degree>::solve()
+  {
+    using StorageType = PatchStorage<MatrixFree<dim, double>>;
+
+    // Get the number of levels in the multigrid hierarchy.
+    const unsigned max_lvl = triangulation.n_global_levels() - 1;
+
+    // Initialize the transfer operator between multigrid levels.
+    MGTransferMatrixFree<dim, LevelNumber> mg_transfer(mg_constrained_dofs);
+    mg_transfer.build(dof_handler);
+
+    // Define the smoother type based on the LaplacePatchSmoother.
+    using SmootherType =
+      Operators::LaplacePatchSmoother<dim, fe_degree, double>;
```

Comment:
This is very nice how compact it can be made here with only a few lines below. Well done!

Reply:
Thank you for noticing!
I am really satisfied with the user interface I was able to design an implement.